src/main/java/io/cdap/plugin/gcp/common/GCPErrorDetailsProvider.java

```diff
@@ -93,6 +103,52 @@ private ProgramFailureException getProgramFailureException(HttpResponseException
       pair.getErrorType(), true, e);
   }
 
+
+  /**
+   * Get a ProgramFailureException with the given error
+   * information from {@link IllegalArgumentException}.
+   *
+   * @param e The IllegalArgumentException to get the error information from.
+   * @return A ProgramFailureException with the given error information.
+   */
+  private ProgramFailureException getProgramFailureException(IllegalArgumentException e, ErrorContext errorContext) {
+    String errorMessage = e.getMessage();
+    String subCategory = "IllegalArgument";
+    String errorMessageFormat = "Error occurred in the phase: '%s'. Error message: %s";
+    return ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategoryEnum.PLUGIN, subCategory), errorMessage,
+      String.format(errorMessageFormat, errorContext.getPhase(), errorMessage), ErrorType.USER, true, e);
+  }
+
+  /**
+   * Get a ProgramFailureException with the given error
+   * information from {@link IllegalStateException}.
+   *
+   * @param e The IllegalStateException to get the error information from.
+   * @return A ProgramFailureException with the given error information.
+   */
+  private ProgramFailureException getProgramFailureException(IllegalStateException e, ErrorContext errorContext) {
+    String errorMessage = e.getMessage();
+    String subCategory = "IllegalState";
+    String errorMessageFormat = "Error occurred in the phase: '%s'. Error message: %s";
+    return ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategoryEnum.PLUGIN, subCategory), errorMessage,
+      String.format(errorMessageFormat, errorContext.getPhase(), errorMessage), ErrorType.USER, true, e);
+  }
+
+  /**
+   * Get a ProgramFailureException with the given error
+   * information from {@link IOException}.
+   *
+   * @param e The IOException to get the error information from.
+   * @return A ProgramFailureException with the given error information.
+   */
+  private ProgramFailureException getProgramFailureException(IOException e, ErrorContext errorContext) {
```

Comment:
How is `IOException` is a `user` error? Even the error messages are not at all actionable.

Reply:
We need to convert only those exceptions to `ProgramFailureException` where we can get enough information from the exceptions.

Reply:
In underlying places, if we are throwing `IOException` we should change it to throw `ProgramFailureException` rather than assuming `IOException` will always be `user` error.

Reply:
There are IOException being thrown from `BigQueryOutputFormat` in code eg.

https://github.com/data-integrations/google-cloud/blob/601f62bd0005255e75bff19db8711a8898358d49/src/main/java/io/cdap/plugin/gcp/bigquery/sink/BigQueryOutputFormat.java#L168

Reply:
how do we convert those to ProgramFailureException ?

Reply:
> In underlying places, if we are throwing `IOException` we should change it to throw `ProgramFailureException` rather than assuming `IOException` will always be `user` error.

Ok got it, wll do these changes.

Reply:
Changed underlying `IOException` to  `ProgramFailureException`.